sycl/include/sycl/detail/spirv.hpp

```diff
@@ -781,217 +809,315 @@ inline uint32_t membermask() {
 }
 #endif
 
+template <typename GroupT>
+inline uint32_t MapShuffleID(GroupT g, id<1> local_id) {
+  if constexpr (is_tangle_or_opportunistic_group<GroupT>::value ||
+                is_ballot_group<GroupT>::value)
+    return detail::IdToMaskPosition(g, local_id);
+  else if constexpr (is_fixed_size_group<GroupT>::value)
+    return g.get_group_linear_id() * g.get_local_range().size() + local_id;
+  else
+    return local_id.get(0);
+}
+
 // Forward declarations for template overloadings
-template <typename T>
-EnableIfBitcastShuffle<T> SubgroupShuffle(T x, id<1> local_id);
+template <typename GroupT, typename T>
+EnableIfBitcastShuffle<T> Shuffle(GroupT g, T x, id<1> local_id);
 
-template <typename T>
-EnableIfBitcastShuffle<T> SubgroupShuffleXor(T x, id<1> local_id);
+template <typename GroupT, typename T>
+EnableIfBitcastShuffle<T> ShuffleXor(GroupT g, T x, id<1> local_id);
 
-template <typename T>
-EnableIfBitcastShuffle<T> SubgroupShuffleDown(T x, uint32_t delta);
+template <typename GroupT, typename T>
+EnableIfBitcastShuffle<T> ShuffleDown(GroupT g, T x, uint32_t delta);
 
-template <typename T>
-EnableIfBitcastShuffle<T> SubgroupShuffleUp(T x, uint32_t delta);
+template <typename GroupT, typename T>
+EnableIfBitcastShuffle<T> ShuffleUp(GroupT g, T x, uint32_t delta);
 
-template <typename T>
-EnableIfGenericShuffle<T> SubgroupShuffle(T x, id<1> local_id);
+template <typename GroupT, typename T>
+EnableIfGenericShuffle<T> Shuffle(GroupT g, T x, id<1> local_id);
 
-template <typename T>
-EnableIfGenericShuffle<T> SubgroupShuffleXor(T x, id<1> local_id);
+template <typename GroupT, typename T>
+EnableIfGenericShuffle<T> ShuffleXor(GroupT g, T x, id<1> local_id);
 
-template <typename T>
-EnableIfGenericShuffle<T> SubgroupShuffleDown(T x, uint32_t delta);
+template <typename GroupT, typename T>
+EnableIfGenericShuffle<T> ShuffleDown(GroupT g, T x, uint32_t delta);
 
-template <typename T>
-EnableIfGenericShuffle<T> SubgroupShuffleUp(T x, uint32_t delta);
+template <typename GroupT, typename T>
+EnableIfGenericShuffle<T> ShuffleUp(GroupT g, T x, uint32_t delta);
 
-template <typename T>
-EnableIfNativeShuffle<T> SubgroupShuffle(T x, id<1> local_id) {
+template <typename GroupT, typename T>
+EnableIfNativeShuffle<T> Shuffle(GroupT g, T x, id<1> local_id) {
+  uint32_t LocalId = MapShuffleID(g, local_id);
 #ifndef __NVPTX__
-  return __spirv_SubgroupShuffleINTEL(convertToOpenCLType(x),
-                                      static_cast<uint32_t>(local_id.get(0)));
+  std::ignore = g;
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT> &&
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
```

Comment:
If the bug is in open-source IGC, perhaps we could provide a link to the corresponding Github issue.

Reply:
The issue has been reported internally. 👍